KHR_implicit_shapes extension Draft Proposal

eoineoineoin opened the pull request:
Draft of an extension which describes generic, non-renderable shapes suitable for collision detection.

## Known Implementations

[Blender importer/exporter](https://github.com/eoineoineoin/glTF_Physics_Blender_Exporter)
[Babylon.js importer](https://github.com/eoineoineoin/glTF_Physics_Babylon)
[Godot importer](https://github.com/eoineoineoin/glTF_Physics_Godot_Importer)

## Sample Assets

https://github.com/eoineoineoin/glTF_Physics/tree/master/samples

eoineoineoin committed:
Merge of KHR_collision_shapes extension

eoineoineoin committed:
Add ability to skin/morph convex and trimesh shapes

eoineoineoin commented on extensions/2.0/Khronos/KHR_collision_shapes/schema/glTF.KHR_collision_shapes.shape.convex.schema.json:
```diff
@@ -0,0 +1,30 @@
+    "$schema": "http://json-schema.org/draft-04/schema",
+    "title": "KHR_collision_shapes Convex Shape",
+    "type": "object",
+    "description": "Parameters describing a convex shape.",
```
Only concern here is that there's slightly different requirements for the `mode` of the `mesh.primitive` objects in the referenced mesh. I'd expect trimesh objects to require `TRIANGLES`, `TRIANGLE_STRIP` or `TRIANGLE_FAN`, though a convex hull could use any `mode`.

If we do want to combine them, would you agree that the `isConvex` flag should *not* be a mutable property?

lexaknyazev replied:
Please add explicit trimesh reqs to the extension spec and to the `mesh` property schema description (also change "ID" to "index" there). The only current mention seems to be the word "triangulated" in the trimesh description, which is not strict enough.

Surely, the flag should be read only.

eoineoineoin committed:
Feedback from review.

eoineoineoin committed:
Convert some minimum properties to exclusiveMinimum, to avoid some degenerate geometries

eoineoineoin commented on extensions/2.0/Khronos/KHR_collision_shapes/schema/glTF.KHR_collision_shapes.shape.box.schema.json:
```diff
@@ -0,0 +1,22 @@
+            "description": "The extents of the box in each axis in local space.",
+            "items": {
+                "type": "number",
+                "minimum": 0.0
```
Yeah, though engines tend to support this, it's probably not a useful construct. I've converted properties to exclusiveMinimum where it makes sense; though left the radii properties on capsules/cylinders alone - consider a cone, which has one radius of 0.

lexaknyazev replied:
Should it be disallowed to have both radii as zeros? Although it's not easily representable with schemas, the spec may have such normative language.

eoineoineoin replied:
OK, I've added language clarifying degenerate shapes (https://github.com/KhronosGroup/glTF/pull/2370/commits/ef428bc842ca14bad7c64426aba7f389d54390f1) - one concern I have though, is how this requirement interacts with node scale; in particular, the third "Implementation Note" in section https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#transformations which allows nodes to be "optimized away" if they have zero scale.

Similarly, I'd _like_ if negative node scales were supported. i.e. a sphere attached to a node with scale [-1, -1, -1] should be identical to the sphere attached to a node with a scale of [1, 1, 1].

Have you any objection if we were to add text describing how to handle those cases?

eoineoineoin committed:
Merge `convex` and `trimesh` objects

eoineoineoin committed:
Add language prohibiting degenerate shapes

lexaknyazev commented on extensions/2.0/Khronos/KHR_collision_shapes/schema/glTF.KHR_collision_shapes.shape.mesh.schema.json:
```diff
@@ -0,0 +1,36 @@
+        "extras": { }
+    },
+    "required": [
+        "convex",
```
Not a strong opinion but would it be reasonable to make the property not required with the `false` being the default value?

lexaknyazev commented on extensions/2.0/Khronos/KHR_collision_shapes/README.md:
```diff
@@ -0,0 +1,134 @@
+
+As both the `mesh` shape references a `mesh`, it additionally allows for optional `skin` and `weights` properties, which have the same semantics and requirements enforced by the properties of the same name associated with a `node`. When specified on a `mesh` whose `convex` property is `true`, the resulting collision shape should be the convex hull of the deformed mesh. As collision detection is t ...
+
+Degenerate shapes are prohibited. A `sphere` must have a positive, non-zero radius. `box` shapes must have positive non-zero values for each component of `size`. `cylinder` and `capsule` shapes must have a positive, non-zero `height` and both `radiusTop` and `radiusBottom` should be positive; at least one of the radii should be non-zero. For `mesh` shapes whose `convex` property is `false`, the r ...
```
Please do not use backticks unless explicitly referring to properties. For example:
> A sphere must...

> Box shapes must...

EricGriffith commented:
1. It would be great to include example usage of the schemas in the spec.

2. **both** seems to be a typo here:
> As both the `mesh` shape

3. The `convex` property name is confusing. To me, it implies that the mesh is required to be convex. I prefer the previous version where convex and mesh were separate shapes. If momentum means it they must stay combined, please use a more descriptive name like `convexHull`.

4. The `skin` and `weights` properties should probably be disallowed when a mesh shape is used together with a node that uses the same mesh for rendering. Otherwise, every extension using these shapes will need to repeat the same normative text explaining the precedence of the two (node + shape) skin properties and the three (node + mesh + shape) weights properties.

eoineoineoin committed:
Various pieces of feedback:

Rename "convex"->"convexHull". Tidy up phrasing, documentation.
Improved introduction text; add an inline sample defining a shape.
Clarifications around degenerate shapes & node scale.
Remove backticks when not referring to a property.

aaronfranke commented on extensions/2.0/Khronos/KHR_collision_shapes/schema/glTF.KHR_collision_shapes.shape.mesh.schema.json:
```diff
@@ -0,0 +1,36 @@
+            "description": "The index of the mesh from which to build the collision representation.",
+            "allOf": [{"$ref": "glTFid.schema.json"}]
+        },
+        "skin": {
+            "description": "The index of a skin with which to deform the mesh.",
+            "allOf": [ { "$ref": "glTFid.schema.json" } ]
+        },
+        "weights": {
+            "type": "array",
+            "description": "The weights of the instantiated morph target.",
+            "minItems": 1,
+            "items": {
+                "type": "number"
+            }
+        },
```
I'm confused, why are skins and weights used for collision meshes? Where was the discussion on this? What is the use case?

eoineoineoin replied:
This was discussed in several Khronos meetings on the interactivity spec. These parameters match functionality already in the glTF node. The usecase is to allow users to describe collision shapes which accurately match the render meshes.

lyuma replied:
> collision shapes which accurately match the render meshes

I could see weights/skins being calculated once at load time, but since "interactivity" is brought up, it sounds like the intent is to permit changing vertices on any given frame. I cannot find a good specification for the intended behavior when vertices of a collision mesh are animated at runtime (e.g. on every frame). I see two big problems:

1. Firstly, it seems highly unperformant: both the need to regenerate optimized collision geometry (such as a convex hull or deduplicated trimesh vertices); and to run skinning on the CPU (some engines upload skinned mesh geometry to the GPU and do not store a copy on the CPU, so this may involve data transfer from the GPU on each frame). In the case of convex hulls, it might not be feasible to re-build the hull every frame in which it may have been animated.

2. In a physics sense, it seems to me problematic that the skinned mesh deformation is non-physical: if the change in shape causes a collision, does it exert a force? if so, what force? if not, will it be allowed to intersect with adjacent geometry from one frame to the next, or entirely phase through it?

As a concrete example, imagine a razor-thin mesh floor of an elevator, using a skin joint to move upwards, and a small marble resting on the floor of this elevator. Imagine an implementation performs CPU deformation by calling BakeMesh() every frame, effectively swapping out one mesh for another. If the joint node moves quickly enough, the floor of the elevator may move more in one physics frame than the radius of the marble, causing it to fall. This will be highly dependent on physics framerate, which could differ either due to hardware specs or by user agent.

Ultimately, I believe the likely outcome is the majority of conventional game engines will choose between ignoring support for skins/morphs for physics shapes (which means inconsistent support), or doing something unperformant (for example, calling BakeMesh() every frame). This sort of feature sounds cool, but it's a pie in the sky, and the likelihood of inconsistent support could scare users away from using the feature, which tells me that it belongs in a separate extension.

eoineoineoin replied:
Note that `mesh` objects already have `weights` defined in the core glTF specification. Those weights may be animated, so this needs to be supported anyway. These properties merely the shapes consistent with how nodes can apply per-instance weights to a mesh, as well as using a skinned mesh.

As with many glTF features, it's possible to describe scenarios which are non-physical or may perform poorly, but at the end of the day, glTF content is made by a human who is ultimately making decisions about what that content is. The fact that a feature _may_ be misused is not a good reason to say "we don't support this feature which is in the core spec."

aaronfranke reviewed (commented):
The contents of this PR do not match the repo in which the extension was developed: https://github.com/eoineoineoin/glTF_Physics/tree/master/extensions/2.0/Khronos/KHR_collision_shapes/schema

If changes are needed, the original repo should also be updated so that it stays in sync.

npolys commented:
note that in X3D

Collision for navigation can be a property on Grouping Node types

https://web3d.org/documents/specifications/19775-1/V4.0/Part01/components/navigation.html#Collision

and then there is a whole Rigid Body Physics Component for other object-object collisions
https://web3d.org/documents/specifications/19775-1/V4.0/Part01/components/rigidBodyPhysics.html
________________________________
From: Aaron Franke ***@***.***>
Sent: Thursday, May 30, 2024 6:17 PM
To: KhronosGroup/glTF ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [KhronosGroup/glTF] Merge of KHR_collision_shapes extension (PR #2370)


@aaronfranke commented on this pull request.

The contents of this PR do not match the repo in which the extension was developed: https://github.com/eoineoineoin/glTF_Physics/tree/master/extensions/2.0/Khronos/KHR_collision_shapes/schema

________________________________

In extensions/2.0/Khronos/KHR_collision_shapes/schema/glTF.KHR_collision_shapes.shape.mesh.schema.json<https://github.com/KhronosGroup/glTF/pull/2370#discussion_r1621478845>:

> +        "skin": {
+            "description": "The index of a skin with which to deform the mesh.",
+            "allOf": [ { "$ref": "glTFid.schema.json" } ]
+        },
+        "weights": {
+            "type": "array",
+            "description": "The weights of the instantiated morph target.",
+            "minItems": 1,
+            "items": {
+                "type": "number"
+            }
+        },


I'm confused, why are skins and weights used for collision meshes? Where was the discussion on this? What is the use case?

—
Reply to this email directly, view it on GitHub<https://github.com/KhronosGroup/glTF/pull/2370#pullrequestreview-2089452159>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB2TSM72PNKRZZVZHCH2ZF3ZE6QOZAVCNFSM6AAAAABE5T5AW6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOBZGQ2TEMJVHE>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

eoineoineoin committed:
Merge latest set of changes from development repository.

    Remove convex hull options from base KHR_collision_shapes; for many
    cases, convex hulls are not necessary and having to support convex hull
    generation in a viewer may be a significant implementation burden.

    Refactor morph targets to support the common use-case of describing a
    mesh collider which intends to use the same morph target weights as a
    node's instance weights. Add an explicit recommendation for how an
    extension can reference a collision shape.

    Upgrade schemas to draft 2020-12.

eoineoineoin committed:
Remove unnecessary known implementations section

wallabyway commented:
@eoineoineoin - Thanks for bringing up 'implicit shapes'.

If I could pick two shapes that would help AEC, it would be cylinder and cube.  We could add a million other shapes, but I would compromise everything, just to get these two into a future spec somehow.

We pay a very high price for cylinders (HVAC, railings, cables, rebar, etc).  If the industry agreed on this, there would be a lot of precision gained, less data-loss.

Same for 'Cubes'.  If the industry agreed on a cube shape, rather than "I think these 16 triangles are a cube", adds a lot of meaning for AEC (design intent, manifold, volume, outline-style, improved z-fighting).

I know it's a very low bar, but you got to start somewhere and these two additions add an impressive amount of value IMHO.

javagl commented:
@wallabyway I've lost track of some of the discussion/developments here, and am (unfortuantely) not up to date. But a while ago, I also advocated for basically creating `KHR_shapes` and `KHR_collision` (using `KHR_shapes`) extensions. Somewhere around https://github.com/eoineoineoin/glTF_Physics/issues/1#issuecomment-1325503582 . The use-case would be for CAD/AEC (maybe going so far that these could even serve as a basis for some CSG at some point...). 

When "shapes" are defined here (non-renderable, as part of a "collision" extension), then someone **will** create a new extension that defines _the same shapes_, but "renderable". And in the worst case, one defines a "box" as (min, max), and the other one defines a "box" as (origin, size), and we all know what happens then...

Some further related discussion is in https://github.com/KhronosGroup/glTF/issues/1135 and around https://github.com/KhronosGroup/glTF/issues/1986#issuecomment-1023290360 , and many issues that are linked to/from these. (It's hard to wrap this up - that's why I lost track...).

eoineoineoin commented:
> @wallabyway I've lost track of some of the discussion/developments here, and am (unfortuantely) not up to date. But a while ago, I also advocated for basically creating `KHR_shapes` and `KHR_collision` (using `KHR_shapes`) extensions. Somewhere around [eoineoineoin/glTF_Physics#1 (comment)](https://github.com/eoineoineoin/glTF_Physics/issues/1#issuecomment-1325503582) . The use-case would be for CAD/AEC (maybe going so far that these could even serve as a basis for some CSG at some point...).
> 
> When "shapes" are defined here (non-renderable, as part of a "collision" extension), then someone **will** create a new extension that defines _the same shapes_, but "renderable". And in the worst case, one defines a "box" as (min, max), and the other one defines a "box" as (origin, size), and we all know what happens then...
> 
> Some further related discussion is in #1135 and around [#1986 (comment)](https://github.com/KhronosGroup/glTF/issues/1986#issuecomment-1023290360) , and many issues that are linked to/from these. (It's hard to wrap this up - that's why I lost track...).

Hey Marco,

Right, it does seem that there's interest in using the shapes for rendering, amongst other things. So, at the very least, there's a rename of the extension and associated reworking of documentation in the works, to ensure it's obviously usable for multiple different use-cases.

The "mesh" shape type is causing us some hassle, though, so I want to see if you have any thoughts here. For things like physics, selectability, UI, etc., it's useful to be able to describe a shape which is derived from a `glTF.mesh.` Though, for the "shape rendering" extension, this `shape.mesh` is redundant, since glTF meshes are already renderable via nodes. That leaves us with a couple of options:

1. Remove `shapes.mesh`, require "physics extension" and friends to reference `glTF.mesh` via some other mechanism
2. Leave as-is; "shape rendering extension" and friends would define how those `shapes.mesh `are handled

My concern is that with option 1, the options available in the `shapes.mesh` today will have to be re-defined by any extension which wants to use meshes. Similarly, with option 2, extensions might all require documentation to say how each shape type is handled (though, this might need to be the case anyway?) Any preferences? Other ideas?

eoineoineoin committed:
Merge changes from refactor of KHR_collision_shapes to KHR_implicit_shapes

eoineoineoin committed:
Feedback, rewording on cones/cylinders

lyuma commented:
I have been reading through the recent changes to use a collider or trigger "geometry" object.

I have some concern with the new version of the spec because it references "node", that implies referencing a node, which must be visible in core gltf. Is there a way to use meshes in physics colliders without rendering those meshes?

Related to this problem is a confusing paragraph here, so I would like some clarification on this:

> Collision shapes are parameterized in local space of the `node` they are associated with. If a shape is required to have an offset from the local space of the node the shape is associated with (for example a sphere _not_ centered at local origin or a rotated box,) a child node should be added with the desired offset applied, and the shape properties added to that child.

if I understand correctly, does this paragraph that the nodes and **child nodes** (a comment by Khronos members implies this actually means descendants) in the context of mesh geometry are taken in local space of the referenced `geometry` / `node`?

if so, can I ask if an acceptable way to implement hidden mesh collision shapes is to make a node whose scale is [0,0,0] (which hides the mesh as per the core gltf spec): since the collision shape children use a coordinate space relative to this node, child nodes will ignore the 0 scale. Does this make any sense?

if for some reason using [0,0,0] scale is not possible, the only alternatives I can think of are using a translation at 1e+99 to effectively hide the mesh, or relying on the node visbility extension (but this will not degrade well to older versions of Blender, for example)

eoineoineoin committed:
Clarifications on "type"

HexaField commented:
@lyuma I can see some potential problems with your suggestions, most importantly with very large translations throwing off bounding volume calculations.

What if the spec allowed for referencing other scenes in the same document? This way, another scene would not be rendered (or optionally rendered, such as for rendering colliders in an editor) yet still accessible and loaded into memory under a separate hierarchy.

eoineoineoin committed:
Merge 35cc5dd4746b43bd from origin repo

eoineoineoin committed:
Add explicit links to every schema

eoineoineoin committed:
Bump status to Release Candidate

aaronfranke commented on extensions/2.0/Khronos/KHR_implicit_shapes/README.md:
```diff
@@ -0,0 +1,140 @@
+
+| |Type|Description|
+|-|-|-|
+|**sphere**|`object`|A sphere centered at the origin in local space.|
```
Nitpick: Tables should be formatted.

aaronfranke commented on extensions/2.0/Khronos/KHR_implicit_shapes/README.md:
```diff
@@ -0,0 +1,140 @@
+* Sphere: [glTF.KHR_implicit_shapes.shape.sphere.schema.json](schema/glTF.KHR_implicit_shapes.shape.sphere.schema.json)
+* Box: [glTF.KHR_implicit_shapes.shape.box.schema.json](schema/glTF.KHR_implicit_shapes.shape.box.schema.json)
+* Cylinder: [glTF.KHR_implicit_shapes.shape.cylinder.schema.json](schema/glTF.KHR_implicit_shapes.shape.cylinder.schema.json)
+* Capsule: [glTF.KHR_implicit_shapes.shape.capsule.schema.json](schema/glTF.KHR_implicit_shapes.shape.capsule.schema.json)
```
Nitpick: Khronos's preferred Markdown list style is using `-` symbols.

eoineoineoin committed:
Trivial formatting changes; output not affected

eoineoineoin committed:
Add planes to KHR_implicit_shapes

aaronfranke commented on extensions/2.0/Khronos/KHR_implicit_shapes/README.md:
```diff
@@ -0,0 +1,147 @@
+
+When the `type` of a shape refers to a shape type defined by this extension, the shape must not supply a value for a different shape type e.g. a shape whose `type` is "box" must not have a defined `sphere` property. Similarly, if `type` does not reference a shape type declared by `KHR_implicit_shapes`, an extension or other mechanism should supply an alternate shape definition.
+
+By default, planes have infinite extents along their tangent and bitangent vectors, specified as X and Z in local space, respectively. A plane may optionally provide `sizeX` and `sizeZ` properties to specify a finite size along these axes.
```
This text is unclear, is the size "along these axes" referring to the radius or the diameter? Also, it should clarify +X and +Z for the tangent vectors.

eoineoineoin committed:
Minor clarifications

lexaknyazev commented on extensions/2.0/Khronos/KHR_implicit_shapes/README.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+To describe the geometry which represents the object, shapes must define at most one of the following properties:
````
Consider requiring the corresponding object (so it's always available in the Object Model).

eoineoineoin replied:
Do you mean that even if a shape was utilizing the default values, you'd want to require an empty object? e.g.

```
"shapes" = [
{
     "type": "sphere",
     "sphere": {}
}
...
```

One complication on REQUIREing one of the objects that other extensions may add additional types and may not wish to have a fallback:

```
"shapes" = [
{
     "type": "teapot",
     "extensions": {
            "ext_teapot_shape" : {
                       ....
            }
    }
}
...
```

For those types, we wouldn't want to require an unrelated object in the shape.

lexaknyazev replied:
Even when an object is syntactically empty it's not semantically empty because most shape properties have default values so, e.g., `"sphere": {}` actually means `"sphere": { "radius": 0.5 }`.

The extension should require an object matching the type so future extensions wouldn't require unrelated objects.

lexaknyazev commented on extensions/2.0/Khronos/KHR_implicit_shapes/README.md:
```diff
@@ -0,0 +1,147 @@
+
+By default, planes have infinite extents along their tangent and bitangent vectors, specified as X and Z in local space, respectively. A plane may optionally provide `sizeX` and `sizeZ` properties to specify a finite total size along these axes.
+
+Degenerate shapes are prohibited. If provided, a plane must have positive non-zero `sizeX` and `sizeZ` extents. A sphere must have a positive, non-zero radius. A box shape must have positive non-zero values for each component of `size`. The cylinder and capsule shapes must have a positive, non-zero `height`, both `radiusTop` and `radiusBottom` must be non-negative, and at least one of `radiusTop` ...
```
Ditto about planes.

hybridherbst commented:
I think the "non-renderable" aspect could be made a bit more clear, at first I was confused on why there's no specification around UVs/tangents for these shapes.

(there's precedent for this kind of confusion creeping in with implicit shapes in OpenUSD, where it's not specified how UVs on implicit shapes work, and subsequently texturing a cube in one app and exporting it to another app as implicit shape is not interoperable at all, with people usually not being aware and then being confused.)

aaronfranke commented:
@hybridherbst A textured cube mesh should be represented as a mesh, not an implicit shape. An implicit shape might accompany such a cube for purposes like collision, but the intended visual representation is a mesh.

I believe the intention of renderability is for things like displaying a plain-colored outline in an editor, not runtime.

hybridherbst commented:
I just think that should be mentioned in the README for this extension to be less confusing. Reading just the extension made me wonder what happens when someone uses an implicit shape in for rendering – easily possible to do "accidentally wrong" via a custom extension. Only the first PR message here has the note "non-renderable".

javagl commented:
Throwback to [a comment from 2022](https://github.com/eoineoineoin/glTF_Physics/issues/1#issuecomment-1331001917) when the basics of what has become _this_ proposal have been sorted out:

> In the area of rendering, different questions arise - mainly that of tesselation or UV mapping.

That said: I think that the fact that the _current_ proposal here is focussing on the building block of the shapes themself makes sense, but think that there could, eventually, be an extension that defines a form of UV mapping for the `KHR_implicit_shapes` _only_ for the case that they are supposed to be used for (textured) rendering.
